### ISSUE TITLE
Linking errors with embree (Hidden symbol is referenced by DSO)

### ISSUE THREAD

**Issue by aoabuhatab** [libigl/libigl#1283]
#### Describe your issue

The included headers: #include <igl/embree/line_mesh_intersection.h>, #include <igl/readSTL.h>
Implemented code: (note: same code works when implemented in the example project provided by libigl, after adding embree option and igl::embree in the target link libraries) 

  std::vector<std::vector<double> > V;
  std::vector<std::vector<int> > F;
  std::vector<std::vector<double> > N;
  bool success = igl::readSTL("File_path.stl",V,F,N);
  Eigen::MatrixXd V_1(V.size(),3);
  Eigen::MatrixXi F_1(F.size(),3);
  Eigen::MatrixXd N_1(N.size(),3);
  Eigen::MatrixXd V_1_f(1,3);
  Eigen::MatrixXd N_1_f(1,3);
  V_1_f(0,0)=rNode.X(); V_1_f(0,1)=rNode.Y(); V_1_f(0,2)=rNode.Z();
  N_1_f(0,0)=rRayVector[0]; N_1_f(0,1)=rRayVector[1]; N_1_f(0,2)=rRayVector[2];

  for (int i=0; i<V.size(); ++i){
      
      V_1(i,0)=V[i][0]; V_1(i,1)=V[i][1]; V_1(i,2)=V[i][2];
      //std::cout<<V_1(i,0)<<V_1(i,1)<<V_1(i,2)<<std::endl;
  }
  for (int i=0; i<F.size(); ++i){
      
      F_1(i,0)=F[i][0]; F_1(i,1)=F[i][1]; F_1(i,2)=F[i][2];
      //std::cout<<F_1(i,0)<<F_1(i,1)<<F_1(i,2)<<std::endl;

  }
  for (int i=0; i<N.size(); ++i){
      
      N_1(i,0)=N[i][0]; N_1(i,1)=N[i][1]; N_1(i,2)=N[i][2];
      //std::cout<<N_1(i,0)<<N_1(i,1)<<N_1(i,2)<<std::endl;

  }

  Eigen::MatrixXd R = igl::embree::line_mesh_intersection<Eigen::MatrixXd, Eigen::MatrixXi >(V_1_f, N_1_f, V_1, F_1);
  std::cout<<"Ray Hit !!!!!!!!!!!!!!!!!!"<<R<<std::endl;

Error:

/usr/bin/ld: ../../bin/filterKernelTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/filterKernelTest] Error 1
src/CMakeFiles/filterKernelTest.dir/build.make:102: recipe for target '../bin/filterKernelTest' failed
make[1]: *** [src/CMakeFiles/filterKernelTest.dir/all] Error 2
CMakeFiles/Makefile2:494: recipe for target 'src/CMakeFiles/filterKernelTest.dir/all' failed
make[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: ../../bin/generalTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/generalTest] Error 1
src/CMakeFiles/generalTest.dir/build.make:102: recipe for target '../bin/generalTest' failed
make[1]: *** [src/CMakeFiles/generalTest.dir/all] Error 2
CMakeFiles/Makefile2:418: recipe for target 'src/CMakeFiles/generalTest.dir/all' failed
[ 95%] Linking CXX executable ../../bin/mathMatrixTest
/usr/bin/ld: ../../bin/mathMatrixTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/mathMatrixTest] Error 1
src/CMakeFiles/mathMatrixTest.dir/build.make:102: recipe for target '../bin/mathMatrixTest' failed
make[1]: *** [src/CMakeFiles/mathMatrixTest.dir/all] Error 2
CMakeFiles/Makefile2:456: recipe for target 'src/CMakeFiles/mathMatrixTest.dir/all' failed
[ 97%] Linking CXX executable ../../bin/meshMotionTest
/usr/bin/ld: ../../bin/meshMotionTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/meshMotionTest] Error 1
src/CMakeFiles/meshMotionTest.dir/build.make:102: recipe for target '../bin/meshMotionTest' failed
make[1]: *** [src/CMakeFiles/meshMotionTest.dir/all] Error 2
CMakeFiles/Makefile2:267: recipe for target 'src/CMakeFiles/meshMotionTest.dir/all' failed
[ 98%] Linking CXX executable ../../bin/geometryToolsTest
[100%] Linking CXX executable ../../bin/nodeSearchTest
/usr/bin/ld: ../../bin/nodeSearchTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
/usr/bin/ld: ../../bin/geometryToolsTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/nodeSearchTest] Error 1
src/CMakeFiles/nodeSearchTest.dir/build.make:102: recipe for target '../bin/nodeSearchTest' failed
make[1]: *** [src/CMakeFiles/nodeSearchTest.dir/all] Error 2
CMakeFiles/Makefile2:343: recipe for target 'src/CMakeFiles/nodeSearchTest.dir/all' failed
make[2]: *** [../bin/geometryToolsTest] Error 1
src/CMakeFiles/geometryToolsTest.dir/build.make:102: recipe for target '../bin/geometryToolsTest' failed
make[1]: *** [src/CMakeFiles/geometryToolsTest.dir/all] Error 2
CMakeFiles/Makefile2:305: recipe for target 'src/CMakeFiles/geometryToolsTest.dir/all' failed
make: *** [all] Error 2
Makefile:129: recipe for target 'all' failed
    + An error occured when running make, see build/CompileLog.txt

#### Check all that apply (change to `[x]`)
- [ ] Windows
- [ ] Mac OS X
- [x] Linux
- [ ] I have tried the `dev` branch and the problem persists

See https://libigl.github.io/CONTRIBUTING/#bugreport for more tips.

**Comment by jdumas** [comment]
What version of linux are you one, and which version of GCC are you using? It looks like your compiler is also picking up a system-wide installed version of embree. Can you try removing the package from your system see if it compiles?

**Comment by aoabuhatab** [comment]
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic

gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0

I tried removing everything related to it, but still got the same error.

/usr/bin/ld: ../../bin/generalTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/generalTest] Error 1
src/CMakeFiles/generalTest.dir/build.make:102: recipe for target '../bin/generalTest' failed
make[1]: *** [src/CMakeFiles/generalTest.dir/all] Error 2
CMakeFiles/Makefile2:418: recipe for target 'src/CMakeFiles/generalTest.dir/all' failed
make[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: ../../bin/filterKernelTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/filterKernelTest] Error 1
src/CMakeFiles/filterKernelTest.dir/build.make:102: recipe for target '../bin/filterKernelTest' failed
make[1]: *** [src/CMakeFiles/filterKernelTest.dir/all] Error 2
CMakeFiles/Makefile2:494: recipe for target 'src/CMakeFiles/filterKernelTest.dir/all' failed
[ 95%] Linking CXX executable ../../bin/mathMatrixTest
/usr/bin/ld: ../../bin/mathMatrixTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/mathMatrixTest] Error 1
src/CMakeFiles/mathMatrixTest.dir/build.make:102: recipe for target '../bin/mathMatrixTest' failed
make[1]: *** [src/CMakeFiles/mathMatrixTest.dir/all] Error 2
CMakeFiles/Makefile2:456: recipe for target 'src/CMakeFiles/mathMatrixTest.dir/all' failed
[ 97%] Linking CXX executable ../../bin/meshMotionTest
/usr/bin/ld: ../../bin/meshMotionTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/meshMotionTest] Error 1
src/CMakeFiles/meshMotionTest.dir/build.make:102: recipe for target '../bin/meshMotionTest' failed
make[1]: *** [src/CMakeFiles/meshMotionTest.dir/all] Error 2
CMakeFiles/Makefile2:267: recipe for target 'src/CMakeFiles/meshMotionTest.dir/all' failed
[ 98%] Linking CXX executable ../../bin/geometryToolsTest
[100%] Linking CXX executable ../../bin/nodeSearchTest
/usr/bin/ld: ../../bin/geometryToolsTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/geometryToolsTest] Error 1
src/CMakeFiles/geometryToolsTest.dir/build.make:102: recipe for target '../bin/geometryToolsTest' failed
make[1]: *** [src/CMakeFiles/geometryToolsTest.dir/all] Error 2
CMakeFiles/Makefile2:305: recipe for target 'src/CMakeFiles/geometryToolsTest.dir/all' failed
/usr/bin/ld: ../../bin/nodeSearchTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/nodeSearchTest] Error 1
src/CMakeFiles/nodeSearchTest.dir/build.make:102: recipe for target '../bin/nodeSearchTest' failed
make[1]: *** [src/CMakeFiles/nodeSearchTest.dir/all] Error 2
CMakeFiles/Makefile2:343: recipe for target 'src/CMakeFiles/nodeSearchTest.dir/all' failed
make: *** [all] Error 2
Makefile:129: recipe for target 'all' failed
    + An error occured when running make, see build/CompileLog.txt

**Comment by jdumas** [comment]
Have you tried the `dev` branch as well?

**Comment by aoabuhatab** [comment]
with dev also same error again.

/usr/bin/ld: ../../bin/filterKernelTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/filterKernelTest] Error 1
src/CMakeFiles/filterKernelTest.dir/build.make:102: recipe for target '../bin/filterKernelTest' failed
make[1]: *** [src/CMakeFiles/filterKernelTest.dir/all] Error 2
CMakeFiles/Makefile2:494: recipe for target 'src/CMakeFiles/filterKernelTest.dir/all' failed
make[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: ../../bin/generalTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/generalTest] Error 1
make[1]: *** [src/CMakeFiles/generalTest.dir/all] Error 2
src/CMakeFiles/generalTest.dir/build.make:102: recipe for target '../bin/generalTest' failed
CMakeFiles/Makefile2:418: recipe for target 'src/CMakeFiles/generalTest.dir/all' failed
[ 95%] Linking CXX executable ../../bin/mathMatrixTest
/usr/bin/ld: ../../bin/mathMatrixTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/mathMatrixTest] Error 1
src/CMakeFiles/mathMatrixTest.dir/build.make:102: recipe for target '../bin/mathMatrixTest' failed
make[1]: *** [src/CMakeFiles/mathMatrixTest.dir/all] Error 2
CMakeFiles/Makefile2:456: recipe for target 'src/CMakeFiles/mathMatrixTest.dir/all' failed
[ 97%] Linking CXX executable ../../bin/nodeSearchTest
[ 98%] Linking CXX executable ../../bin/meshMotionTest
/usr/bin/ld: ../../bin/meshMotionTest: hidden symbol/ `usrrtcSetGeometryMask/'bin /inld :/ home./.aoabuhatab//.Projects.//HiWibin//optimizationFrameWorknodeSearchTest/:thirdParty /hiddenlib /symbollibembree3.a (`rtcore.cpp.ortcSetGeometryMask)'  isin  referenced/ homeby/ aoabuhatabDSO/
Projects//usrHiWi//binoptimizationFrameWork//ldthirdParty:/ libfinal/ libembree3.alink( rtcore.cpp.ofailed):  isBad  referencedvalue 
by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/nodeSearchTest] Error 1
src/CMakeFiles/nodeSearchTest.dir/build.make:102: recipe for target '../bin/nodeSearchTest' failed
make[1]: *** [src/CMakeFiles/nodeSearchTest.dir/all] Error 2
CMakeFiles/Makefile2:343: recipe for target 'src/CMakeFiles/nodeSearchTest.dir/all' failed
make[2]: *** [../bin/meshMotionTest] Error 1
src/CMakeFiles/meshMotionTest.dir/build.make:102: recipe for target '../bin/meshMotionTest' failed
make[1]: *** [src/CMakeFiles/meshMotionTest.dir/all] Error 2
CMakeFiles/Makefile2:267: recipe for target 'src/CMakeFiles/meshMotionTest.dir/all' failed
[100%] Linking CXX executable ../../bin/geometryToolsTest
/usr/bin/ld: ../../bin/geometryToolsTest: hidden symbol `rtcSetGeometryMask' in /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/lib/libembree3.a(rtcore.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/geometryToolsTest] Error 1
src/CMakeFiles/geometryToolsTest.dir/build.make:102: recipe for target '../bin/geometryToolsTest' failed
make[1]: *** [src/CMakeFiles/geometryToolsTest.dir/all] Error 2
CMakeFiles/Makefile2:305: recipe for target 'src/CMakeFiles/geometryToolsTest.dir/all' failed
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
    + An error occured when running make, see build/CompileLog.txt

**Comment by aoabuhatab** [comment]
I get this before the error occurs. I don't think I got it before.  

In file included from /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/libigl/cmake/../external/embree/include/embree3/rtcore.h:19:0,
                 from /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/libigl/include/igl/embree/EmbreeIntersector.h:24,
                 from /home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/libigl/include/igl/embree/reorient_facets_raycast.cpp:13:
/home/aoabuhatab/Projects/HiWi/optimizationFrameWork/thirdParty/libigl/cmake/../external/embree/include/embree3/rtcore_version.h:23:0: warning: "EMBREE_STATIC_LIB" redefined
 #define EMBREE_STATIC_LIB
 
<command-line>:0:0: note: this is the location of the previous definition

**Comment by aoabuhatab** [comment]
If I comment out: target_compile_definitions(igl_embree ${IGL_SCOPE} -DEMBREE_STATIC_LIB)
in libigl.cmake, the previous warning doesn't appear, but the hidden symbol error still remains.

**Comment by BruegelN** [comment]
I've made a MWE based on libigl-example-project: https://github.com/BruegelN/libigl-example-project/tree/libigl-issue-1283
I had to add missing data types and commented two lines because `rNode` and `rRayVector` isn't defined. But this should not be related to your issue.
This compiles just fine on my machine(Ubuntu 18.04 with gcc 7.4.0) if you put libigl next to it.
```
nico@local:~/workingDir$ ls 
libigl        ligigl-example-project
[...]
```
